File: 0463-island-perimeter/0463-island-perimeter.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
const islandPerimeter = function(grid) {
    const dir = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    let perimeter = 0;

    for (let rowIdx = 0; rowIdx < grid.length; rowIdx++) {
        for (let colIdx = 0; colIdx < grid[rowIdx].length; colIdx++) {
            if (grid[rowIdx][colIdx] === 1) {
                for (let [dr, dy] of dir) {
                    let nr = rowIdx + dr;
                    let ny = colIdx + dy;
                    if (nr < 0 || ny < 0 || nr >= grid.length || ny >= grid[rowIdx].length || grid[nr][ny] === 0) perimeter++;
                }
            }
        }
    }

    return perimeter;

};

File: 0463-island-perimeter/0463-island-perimeter.py
from collections import deque

class Solution(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        N , M = len(grid), len(grid[0])
        q = deque([])
        dir = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        visited = [[False] * M for _ in range(N)]
        perimeter = 0
        found = False

        for r in range(N):
            for c in range(M):
                if grid[r][c] == 1:
                    q.append((r, c))
                    visited[r][c] = True
                    found = True
            if found:
                break        

        while q:
            r, c = q.popleft()

            for dr, dc in dir:
                nr, nc = r + dr, c + dc
                # not in the range or water
                if nr < 0 or nr > N-1 or nc < 0 or nc > M-1 or grid[nr][nc] == 0:
                    perimeter += 1
                # land
                elif not visited[nr][nc] and grid[nr][nc] == 1:    
                    visited[nr][nc] = True
                    q.append((nr, nc))
                
        return perimeter            